### PR TITLE
Added Microsoft Federation

### PR DESCRIPTION
**Pull request by gershom96**
### Related Issues
- Issue `#149`

### Checklist
- [ ] e2e cypress tests locally verified.
- [ ] Manual test round performed and verified.
- [ ] UX/UI review done on the final implementation.
- [ ] Documentation provided. (Add links if there are any)
- [ ] Unit tests provided. (Add links if there are any)
- [ ] Integration tests provided. (Add links if there are any)

### Security checks
- [ ] Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines?
- [ ] Ran FindSecurityBugs plugin and verified report?
- [ ] Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets?

### CODE REVIEW

File: apps/console/src/extensions/configs/attribute.ts [hunk 1]
```diff
@@ -1,7 +1,7 @@
 /**
- * Copyright (c) 2021, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
+ * Copyright (c) 2022, WSO2 LLC. (https://www.wso2.com). All Rights Reserved.
```
Comment (brionmario) [inline apps/console/src/extensions/configs/attribute.ts:2]: Keep the original year.

```suggestion
 * Copyright (c) 2021, WSO2 LLC. (https://www.wso2.com). All Rights Reserved.
```

File: apps/console/src/features/identity-providers/api/identity-provider.ts [hunk 2]
```diff
@@ -73,7 +73,8 @@ export const createIdentityProvider = (identityProvider: object): Promise<any> =
         method: HttpMethods.POST,
         url: store.getState().config.endpoints.identityProviders
     };
-
+    console.log(requestConfig);
```
Comment (brionmario) [inline apps/console/src/features/identity-providers/api/identity-provider.ts:76]:
```suggestion

```

File: apps/console/src/features/identity-providers/components/forms/authenticators/microsoft-authenticator-form.tsx [hunk 3]
```diff
@@ -0,0 +1,560 @@
+/**
+ * Copyright (c) 2021, WSO2 LLC. (https://www.wso2.com). All Rights Reserved.
```
Comment (brionmario) [inline apps/console/src/features/identity-providers/components/forms/authenticators/microsoft-authenticator-form.tsx:2]:
```suggestion
 * Copyright (c) 2022, WSO2 LLC. (https://www.wso2.com). All Rights Reserved.
```

File: apps/console/src/features/identity-providers/components/wizards/authenticator-create-wizard-factory.tsx [hunk 4]
```diff
@@ -16,6 +16,7 @@
  * under the License.
  */
 
+// eslint-disable-next-line header/header
```
Comment (brionmario) [inline apps/console/src/features/identity-providers/components/wizards/authenticator-create-wizard-factory.tsx:19]: Don't supress lint warnings.
```suggestion
```

File: apps/console/src/features/applications/components/settings/sign-on-methods/sign-on-methods.tsx [hunk 5]
```diff
@@ -121,6 +122,7 @@ export const SignOnMethods: FunctionComponent<SignOnMethodsPropsInterface> = (
     const [ googleAuthenticators, setGoogleAuthenticators ] = useState<GenericAuthenticatorInterface[]>(undefined);
     const [ gitHubAuthenticators, setGitHubAuthenticators ] = useState<GenericAuthenticatorInterface[]>(undefined);
     const [ facebookAuthenticators, setFacebookAuthenticators ] = useState<GenericAuthenticatorInterface[]>(undefined);
+    // const [ microsoftAuthenticators, setMicrosoftAuthenticators ] = useState<GenericAuthenticatorInterface[]>(undefined);
```
Comment (brionmario) [inline apps/console/src/features/applications/components/settings/sign-on-methods/sign-on-methods.tsx:125]: Let'' not commit any commented code. Add them by checking the git history later if needed.

File: apps/console/src/features/identity-providers/api/identity-provider.ts (hunk 2 above)
Comment (gershom96) [reply apps/console/src/features/identity-providers/api/identity-provider.ts]: resolved

File: apps/console/src/features/applications/components/settings/sign-on-methods/sign-on-methods.tsx (hunk 5 above)
Comment (gershom96) [reply apps/console/src/features/applications/components/settings/sign-on-methods/sign-on-methods.tsx]: resolved

File: apps/console/src/features/core/configs/app.ts [hunk 6]
```diff
@@ -49,13 +49,12 @@ export class Config {
...
      *
      */
-    // eslint-disable-next-line @typescript-eslint/no-empty-function
     private constructor() { }
 
     /**
      * This method adds organization path to the server host if an organization is selected.
      *
-     * @returns string
+     * @returns \{string\}
```
Comment (brionmario) [inline apps/console/src/features/core/configs/app.ts:57]:
```suggestion
     * @returns Server host.
```

**Review by brionmario — APPROVED** [review]

**Review by brionmario — APPROVED** [review]

File: apps/console/src/features/core/configs/app.ts [hunk 7]
```diff
@@ -1,5 +1,5 @@
 /**
- * Copyright (c) 2022, WSO2 LLC. (https://www.wso2.com). All Rights Reserved.
+ * Copyright (c) 2020, WSO2 LLC. (https://www.wso2.com). All Rights Reserved.
```
Comment (gershom96) [inline apps/console/src/features/core/configs/app.ts:2]: Correct this. This shouldnt be 2020.

File: apps/console/src/features/identity-providers/components/wizards/microsoft/microsoft-authentication-provider-create-wizard-content.tsx [hunk 8]
```diff
@@ -0,0 +1,295 @@
...
+    const { t } = useTranslation();
+
+    const [ idpList, setIdPList ] = useState<IdentityProviderListResponseInterface>({});
+    const [ isIdPListRequestLoading, setIdPListRequestLoading ] = useState<boolean>(undefined);
+
+    /**
+     * Loads the identity provider authenticators on initial component load.
+     */
+    useEffect(() => {
+
```
Comment (ThaminduDilshan) [inline apps/console/src/features/identity-providers/components/wizards/microsoft/microsoft-authentication-provider-create-wizard-content.tsx:100]: Redundant new line

File: apps/console/src/features/identity-providers/components/wizards/microsoft/microsoft-authentication-provider-create-wizard-content.tsx [hunk 9]
```diff
@@ -0,0 +1,295 @@
...
+     */
+    const idpNameValidation = (value): string => {
+
+        let nameExist = false;
+
+        if (idpList?.count > 0) {
+            idpList?.identityProviders.map((idp) => {
+                if (idp?.name === value) {
+                    nameExist = true;
+
```
Comment (ThaminduDilshan) [inline apps/console/src/features/identity-providers/components/wizards/microsoft/microsoft-authentication-provider-create-wizard-content.tsx:137]: Redundant new line

File: apps/console/src/features/identity-providers/components/wizards/microsoft/microsoft-authentication-provider-create-wizard.tsx [hunk 10]
```diff
@@ -0,0 +1,535 @@
...
+                    onIDPCreate();
+                })
+                .catch((error) => {
+
+                    const identityAppsError: IdentityAppsError = identityProviderConfig.useNewConnectionsView
+                        ? IdentityProviderManagementConstants.ERROR_CREATE_LIMIT_REACHED
+                        : IdentityProviderManagementConstants.ERROR_CREATE_LIMIT_REACHED_IDP;
+
+                    if (error.response.status === 403 &&
+                    error?.response?.data?.code ===
```
Comment (ThaminduDilshan) [inline apps/console/src/features/identity-providers/components/wizards/microsoft/microsoft-authentication-provider-create-wizard.tsx:198]: Check indentation. Might fit 2 out of 3 lines in one line

File: apps/console/src/features/identity-providers/components/forms/authenticators/microsoft-authenticator-form.tsx [hunk 11]
```diff
@@ -0,0 +1,560 @@
...
+
+                                        const scopeMeta: ScopeMetaInterface = resolveScopeMetadata(scope);
+
+                                        return (
+                                            <div
+                                                key={ index }
+                                                className="authenticator-dynamic-property"
+                                                data-testid={ scope }
+                                            >
+                                                <div className="authenticator-dynamic-property-name-container">
```
Comment (ThaminduDilshan) [inline apps/console/src/features/identity-providers/components/forms/authenticators/microsoft-authenticator-form.tsx:491]: Check indentation

File: apps/console/src/features/identity-providers/components/wizards/microsoft/microsoft-authentication-provider-create-wizard-content.tsx [hunk 12]
```diff
@@ -0,0 +1,295 @@
...
+     * @returns error msg if name is already taken.
+     */
+    const idpNameValidation = (value): string => {
+
+        let nameExist = false;
+
+        if (idpList?.count > 0) {
+            idpList?.identityProviders.map((idp) => {
+                if (idp?.name === value) {
+                    nameExist = true;
```
[the remark this answers is not in the document]
Comment (gershom96) [reply apps/console/src/features/identity-providers/components/wizards/microsoft/microsoft-authentication-provider-create-wizard-content.tsx:136]: cannot define function this way. IDE throws an error

File: apps/console/src/features/identity-providers/components/wizards/microsoft/microsoft-authentication-provider-create-wizard.tsx (hunk 10 above)
Comment (gershom96) [reply apps/console/src/features/identity-providers/components/wizards/microsoft/microsoft-authentication-provider-create-wizard.tsx:198]: This cannot be done as the line length exceeds 120

File: apps/console/src/features/identity-providers/components/forms/authenticators/microsoft-authenticator-form.tsx (hunk 11 above)
Comment (ThaminduDilshan) [reply apps/console/src/features/identity-providers/components/forms/authenticators/microsoft-authenticator-form.tsx:491]: nvm

**Pull request merged**